File: open-telemetry/course/04-api-sdk-architecture/01-approaches/python/manual_sdk.py
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import ConsoleSpanExporter, SimpleSpanProcessor
from opentelemetry.sdk.resources import Resource

def main():
    print("=== Running with FULL SDK Registered ===")

    # 1. Application configures SDK
    res = Resource.create({"service.name": "order-service"})
    provider = TracerProvider(resource=res)
    provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))
    trace.set_tracer_provider(provider)

    # 2. Same library code as before
    tracer = trace.get_tracer("library-sample", "1.0.0")

    with tracer.start_as_current_span("library_operation") as span:
        print(f"Span is_recording: {span.is_recording()}")
        span.set_attribute("data.processed", 100)
        span.add_event("completed")

        ctx = span.get_span_context()
        print(f"Is valid trace context: {ctx.is_valid}")
        print(f"Trace ID: {trace.format_trace_id(ctx.trace_id)}")

    provider.shutdown()
    print("=== Result: Telemetry actively processed and exported! ===\n")

if __name__ == "__main__":
    main()

File: open-telemetry/course/04-api-sdk-architecture/01-approaches/java/src/main/java/com/course/ManualSdk.java
package com.course;

import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.exporter.logging.LoggingSpanExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor;

public class ManualSdk {

    public static void main(String[] args) {
        System.out.println("=== Running Java with FULL SDK Registered ===");

        Resource resource = Resource.getDefault().merge(
            Resource.create(Attributes.of(AttributeKey.stringKey("service.name"), "order-service"))
        );

        SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
            .setResource(resource)
            .addSpanProcessor(SimpleSpanProcessor.create(LoggingSpanExporter.create()))
            .build();

        OpenTelemetrySdk sdk = OpenTelemetrySdk.builder()
            .setTracerProvider(tracerProvider)
            .buildAndRegisterGlobal();

        Tracer tracer = sdk.getTracer("library-sample", "1.0.0");

        Span span = tracer.spanBuilder("library_operation").startSpan();
        try {
            System.out.println("Span isRecording: " + span.isRecording());
            System.out.println("Is valid span context: " + span.getSpanContext().isValid());
            System.out.println("Trace ID: " + span.getSpanContext().getTraceId());

            span.setAttribute("data.processed", 100L);
            span.addEvent("completed");
        } finally {
            span.end();
        }

        tracerProvider.shutdown();
        System.out.println("=== Result: Telemetry exported to logs! ===\n");
        System.exit(0);
    }
}
